webapp/api/github.py

```diff
@@ -0,0 +1,139 @@
+import flask
+from urllib.parse import urlencode
+from webapp import api
+from requests.exceptions import RequestException
+
+
+class GithubAPIError(RequestException):
+    pass
+
+
+class GithubAPI:
+    """
+    Provides authentication for GitHub users. Helper methods are also provided
+    for checking organization access and getting user data from the Github API.
+    """
+
+    BASE_URL = "https://api.github.com/"
+    AUTH_URL = "https://github.com/login/oauth/"
+
+    def __init__(
+        self, client_id, client_secret, session=api.requests.Session()
+    ):
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self.session = session
+        self.session.headers["Accept"] = "application/json"
+
+    def authorize(self, scope=None, state=None):
+        """
+        Redirect to Github and request access to a user's data.
+        """
+        params = {
+            "client_id": self.client_id,
+        }
+
+        if scope:
+            params["scope"] = scope
+
+        if state:
+            params["state"] = state
+
+        auth_url = f"{self.AUTH_URL}authorize?{urlencode(params)}"
+
+        return flask.redirect(auth_url)
+
+    def verify_auth(self, code):
```

Comment:
If we follow the logic of my point about removing the `authorize` function, I think maybe this function also doesn't belong i this class. The auth flow is a bit of a special case, and this logic again feels to me like it really belongs in the view function.

If you were to also move this logic to `oauth_views.py` then I don't think you'd have any need for the `GithubAPI` class in that file at all which would solve your duplication problem. And then this class would be simply about making authenticated calls to the API to retrieve actual data, which seem neatly encapsulated.

Do you agree?

Reply:
I don't see the problem in having a function to exchange the authorization_code for the access_token of the user, plus you told me before to change the `request` method to `_method`, if we need to use that method in the view this is not consistent.

Reply:
Maybe what I should change is the name of this method?